pkg/datapath/loader/netlink.go

```diff
@@ -69,85 +66,153 @@ func replaceQdisc(ifName string) error {
+			l.Debug("Finalizing bpffs map migration")
+			if err := bpf.FinalizeBPFFSMigration(bpf.MapPrefixPath(), spec, false); err != nil {
+				l.WithError(err).Error("Could not finalize bpffs map migration")
+			}
 		}
+
+		// Retry loading the Collection after starting map migration.
+		coll, err = bpf.LoadCollection(spec, opts)
 	}
+	if err != nil {
+		return nil, fmt.Errorf("error loading eBPF collection into the kernel: %w", err)
+	}
+	defer coll.Close()
 
-	// If the iproute2 call below is successful, any 'pending' map pins will be removed.
-	// If not, any pending maps will be re-pinned back to their initial paths.
-	cmd := exec.CommandContext(ctx, loaderProg, args...).WithFilters(libbpfFixupMsg)
-	if _, err := cmd.CombinedOutput(log, true); err != nil {
-		// Program/object replacement unsuccessful, revert bpffs migration.
-		if err := bpf.FinalizeBPFFSMigration(bpf.MapPrefixPath(), objPath, true); err != nil {
+	if err := attachProgram(link, coll.Programs[progName], directionToParent(direction), xdpModeToFlag(xdpMode)); err != nil {
+		// Program replacement unsuccessful, revert bpffs migration.
+		if err := bpf.FinalizeBPFFSMigration(bpf.MapPrefixPath(), spec, true); err != nil {
 			return nil, fmt.Errorf("Failed to revert bpffs map migration: %w", err)
 		}
-		return nil, fmt.Errorf("Failed to load prog with %s: %w", loaderProg, err)
+
+		return nil, fmt.Errorf("program %s: %w", progName, err)
 	}
 
-	finalize := func() {
-		l := log.WithField("device", ifName).WithField("objPath", objPath)
-		l.Debug("Finalizing bpffs map migration")
-		if err := bpf.FinalizeBPFFSMigration(bpf.MapPrefixPath(), objPath, false); err != nil {
-			l.WithError(err).Error("Could not finalize bpffs map migration")
+	return finalize, nil
+}
+
+// attachProgram attaches prog to link.
+// If xdpFlags is non-zero, attaches prog to XDP.
+func attachProgram(link netlink.Link, prog *ebpf.Program, qdiscParent uint32, xdpFlags uint32) error {
+	if prog == nil {
+		return errors.New("cannot attach a nil program")
+	}
+
+	if xdpFlags != 0 {
+		// Omitting XDP_FLAGS_UPDATE_IF_NOEXIST equals running 'ip' with -force,
+		// and will clobber any existing XDP attachment to the interface.
+		if err := netlink.LinkSetXdpFdWithFlags(link, prog.FD(), int(xdpFlags)); err != nil {
+			return fmt.Errorf("attaching XDP program to interface %s: %w", link.Attrs().Name, err)
 		}
+
+		return nil
 	}
 
-	return finalize, nil
+	if err := replaceQdisc(link); err != nil {
+		return fmt.Errorf("replacing clsact qdisc for interface %s: %w", link.Attrs().Name, err)
+	}
+
+	filter := &netlink.BpfFilter{
+		FilterAttrs: netlink.FilterAttrs{
+			LinkIndex: link.Attrs().Index,
+			Parent:    qdiscParent,
+			Handle:    netlink.MakeHandle(0, 1),
+			Protocol:  unix.ETH_P_ALL,
+			Priority:  uint16(option.Config.TCFilterPriority),
+		},
+		Fd:           prog.FD(),
+		Name:         fmt.Sprintf("cilium-%s", link.Attrs().Name),
+		DirectAction: true,
+	}
+
```

Comment:
This was changed to a u32 to prevent the user from being able to specify negative values and because there's no u16 equivalent of `viper.GetUint32()`.

What's your understanding of truncating vs. limiting? I would suggest input validation against `math.MaxUint16` that makes the agent bail out early. Before this PR, an error would've occurred when trying to call tc during prog attach, while my current proposal truncates silently, which might be surprising.

Reply:
I see. Yes, I like your idea around input validation early in the option package using `math.MaxUint16` and we can even fatal out if the user provides invalid input like that. In this sort of case what I would prefer is to get the value into the correct type as soon as possible so that only the option package needs to worry about this viper "no GetUint16() method" issue. Then everywhere else in the daemon can correctly assume the correct type rather than having casts like this scattered around.